lib/services/service-scaffold.ts

```diff
@@ -0,0 +1,257 @@
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+/* tslint:disable: max-classes-per-file */
+
+import TypedError = require('typed-error');
+import * as Promise from 'bluebird';
+import * as bodyParser from 'body-parser';
+import * as express from 'express';
+import { RequestHandler } from 'express';
+import * as _ from 'lodash';
+import { Worker } from '../framework/worker';
+import { WorkerClient } from '../framework/worker-client';
+import { Logger, LogLevel } from '../utils/logger';
+import {
+	ExpressServerDetails,
+	ServiceScaffoldErrorCode, ServiceScaffoldEvent, ServiceScaffoldWorkerEvent
+} from './service-scaffold-types';
+import {
+	ServiceAPIHandle, ServiceEmitContext, ServiceEmitRequest, ServiceEmitResponse,
+	ServiceEmitter, ServiceListener, ServiceRegistration,
+} from './service-types';
+
+/**
+ * A TypedError that encapsulates the subtype of the error and message provided.
+ */
+export class ServiceScaffoldError extends TypedError {
+	/** What category of thing went wrong. */
+	public code: ServiceScaffoldErrorCode;
+
+	/**
+	 * @param code     Sub-type of the error.
+	 * @param message  The error message, is stored in the standard location.
+	 */
+	constructor(code: ServiceScaffoldErrorCode, message: string) {
+		super(message);
+		this.code = code;
+	}
+}
+
+/**
+ * A scaffold class to handle a bunch of the repetitive work associated with being a ServiceListener and ServiceEmitter
+ *
+ * This class provides:
+ * - An express app, if asked for.
+ * - A logger.
+ * - An event registration and handling standard.
+ * - Checks for and extracts data from context.
+ * - A standard way of getting contextualised workers.
+ *
+ * Any inheriting child must:
+ * - Provide a `emitData` method which takes a ServiceEmitContext and returns a promise.
+ * - Provide a `verify` function which should check that incoming data isn't spoofed.
+ * - Provide a `serviceName` getter, which because it is based on file path cannot be inherited away.
+ * - Provide a `apiHandle` getter, which should return the underlying object that executes the requests.
+ * - Enqueue events with `context` and `event`, as per `ServiceScaffoldEvent`.
+ */
+export abstract class ServiceScaffold<T> extends WorkerClient<T> implements ServiceListener, ServiceEmitter {
+	/** Endpoint listener that may be used to, for example, receive web hooks. */
+	protected expressApp?: express.Express;
+
+	/** A place to put output for debug and reference. */
+	private loggerInstance = new Logger();
+
+	/** Store a list of actions to perform when particular actions happen */
+	private eventListeners: { [event: string]: ServiceRegistration[] } = {};
+
+	// https://github.com/resin-io-modules/resin-procbots/issues/262
+	constructor(ingress?: ExpressServerDetails) {
```

Comment:
This has changed, hasn't it? What happens with ServiceListener's that don't use Express for listening? They're always going to have an Express instance allocated consuming resources that never gets used?

This should be down to the ServiceListener constructor requesting an instance, not one being provided defacto.